libsolidity/ast/Types.cpp

```diff
@@ -3127,6 +3127,12 @@ BoolResult FunctionType::isImplicitlyConvertibleTo(Type const& _convertTo) const
 	if (convertTo.kind() != kind())
 		return BoolResult::err("Special functions cannot be converted to function types.");
 
+	if (
+		kind() == FunctionType::Kind::Declaration &&
+		m_declaration != convertTo.m_declaration
+	)
+		return BoolResult::err("Function declaration types referring to different functions cannot be converted to each other.");
```

Comment:
What about other kinds of functions with non-null `m_declaration`? Is keeping them convertible really okay?

Reply:
Yes :-).

Reply:
Why? I mean long-term. It may not affect the `selector` special case, but it's surely not good to be able to convert from type with one declaration to a different type with a completely different one?

Reply:
These are ``FunctionType::Kind::Declaration`` types, not internal or external functions. They don't refer to something you can actually call, but *just* to the declaration (hence the name).
You *don't* want to be able to convert a reference to a mere declaration ``C.f`` to be convertible to ``C.g``- the only property they have is their signature, but that's *different* between them, causing the issue we're solving here.

That being said, we should double check the behaviour of references to internal base functions (see https://github.com/ethereum/solidity/pull/14365/files#r1243712169).

In particular, the following should keep working:
```
contract C {
  function f() internal {}
  function g() internal {}
}
contract D is C {
  function h(bool c) internal {
    (c ? C.f : C.g)();
  }
}
```

This may relate to https://github.com/ethereum/solidity/pull/14367, which may need adjustment:
In complex expressions you will get the mobile type of a type - we may need to allow ``FunctionKind::Declaration`` to get a mobile type with the declaration pruned - that way this case should continue to work and properly refer to an internal function pointers.
It's a bit of a mess, though, so I'm not entirely sure myself without investigating further (which we should).